### ISSUE TITLE
[BUG] Includes: declaration for newfunc is wrong

### ISSUE THREAD

**Issue by navytux** [cython/cython#4822]
Hello up there. object.pxd currently declares `newfunc` as follows:

```pyx
ctypedef object (*newfunc)(cpython.type.type, object, object)  # (type, args, kwargs)
```

which implies that `args` and `kwargs` are always live objects and cannot be NULL.

However Python can, and does, call tp_new with either args=NULL, or kwargs=NULL or both. And in such cases this leads to segfault in automatically-generated __Pyx_INCREF for args or kw.

The fix is to change `object` to `PyObject*` for both args and kwargs.

Please see below for details:

```cython
# cython: language_level=3
from cpython cimport newfunc, type as cpytype, Py_TYPE

cdef class X:
    cdef int i
    def __init__(self, i):
        self.i = i
    def __repr__(self):
        return 'X(%d)' % self.i

cdef newfunc _orig_tp_new = Py_TYPE(X(0)).tp_new

cdef object _trace_tp_new(cpytype cls, object args, object kw):
    print('_trace_tp_new', cls, args, kw)
    return _orig_tp_new(cls, args, kw)

Py_TYPE(X(0)).tp_new = _trace_tp_new


x = X(123)
print(x)
```

```console
(neo) (py3.venv) (g.env) kirr@deca:~/src/tools/go/pygolang$ cythonize -i x.pyx 
Compiling /home/kirr/src/tools/go/pygolang/x.pyx because it changed.
[1/1] Cythonizing /home/kirr/src/tools/go/pygolang/x.pyx
running build_ext
building 'x' extension
...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -ffile-prefix-map=/build/python3.9-RNBry6/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -ffile-prefix-map=/build/python3.9-RNBry6/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/home/kirr/src/wendelin/venv/py3.venv/include -I/usr/include/python3.9 -c /home/kirr/src/tools/go/pygolang/x.c -o /home/kirr/src/tools/go/pygolang/tmpqkz1r96s/home/kirr/src/tools/go/pygolang/x.o
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 -Wl,-z,relro -g -fwrapv -O2 -g -ffile-prefix-map=/build/python3.9-RNBry6/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 /home/kirr/src/tools/go/pygolang/tmpqkz1r96s/home/kirr/src/tools/go/pygolang/x.o -o /home/kirr/src/tools/go/pygolang/x.cpython-39-x86_64-linux-gnu.so
```

```console
(neo) (py3.venv) (g.env) kirr@deca:~/src/tools/go/pygolang$ python -c 'import x'
Ошибка сегментирования (стек памяти сброшен на диск)
```

```console
(neo) (py3.venv) (g.env) kirr@deca:~/src/tools/go/pygolang$ gdb python core 
...
Reading symbols from python...
Reading symbols from /usr/lib/debug/.build-id/f9/02f8a561c3abdb9c8d8c859d4243bd8c3f928f.debug...
[New LWP 218557]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `python -c import x'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  _Py_INCREF (op=0x0) at /usr/include/python3.9/object.h:408
408         op->ob_refcnt++;

(gdb) bt 5
#0  _Py_INCREF (op=0x0) at /usr/include/python3.9/object.h:408
#1  __pyx_f_1x__trace_tp_new (__pyx_v_cls=0x7f5ce75e6880 <__pyx_type_1x_X>, __pyx_v_args=(123,), __pyx_v_kw=0x0) at /home/kirr/src/tools/go/pygolang/x.c:1986
#2  0x000000000051dd7e in type_call (type=type@entry=0x7f5ce75e6880 <__pyx_type_1x_X>, args=args@entry=(123,), kwds=kwds@entry=0x0)
    at ../Objects/typeobject.c:1014
#3  0x00007f5ce75df8d4 in __Pyx_PyObject_Call (func=<type at remote 0x7f5ce75e6880>, arg=(123,), kw=0x0) at /home/kirr/src/tools/go/pygolang/x.c:3414
#4  0x00007f5ce75df276 in __pyx_pymod_exec_x (__pyx_pyinit_module=<optimized out>) at /home/kirr/src/tools/go/pygolang/x.c:3017
(More stack frames follow...)

(gdb) f 1
#1  __pyx_f_1x__trace_tp_new (__pyx_v_cls=0x7f5ce75e6880 <__pyx_type_1x_X>, __pyx_v_args=(123,), __pyx_v_kw=0x0) at /home/kirr/src/tools/go/pygolang/x.c:1986
1986      __Pyx_INCREF(__pyx_v_kw);
```

**Expected behavior**
No segfault in autogenerated coce if `_trace_tp_new` matches `newfunc` signature.

**Environment (please complete the following information):**
 - OS: Linux
 - Python version: 3.9.2
 - Cython version: 0.29.30

Thanks beforehand,
Kirill

**Comment by da-woods** [comment]
In this case the best solution is probably just to change the pxd file to use PyObject*, as you recommend. PR welcome.

This kind of problem does make me think it'd occasionally be useful to have a "object but may be null" type.

**Comment by navytux** [comment]
@da-woods, thanks for feedback. I've created https://github.com/cython/cython/pull/4823.

**Comment by scoder** [comment]
Hmm. Is this really something we should change in 0.29.x? Isn't there a risk of breaking working code? That shouldn't be done in a point release.

**Comment by da-woods** [comment]
> Hmm. Is this really something we should change in 0.29.x? Isn't there a risk of breaking working code? That shouldn't be done in a point release.

Possibly. The conclusion I came to was that any code that used it was likely to just crash and not work. But feel free to revert and leave it to 3.0 if you disagree.

**Comment by scoder** [comment]
We could argue that if that declaration was unusable in 0.29.x, then probably no-one used it. So it's not worth repairing for legacy code. However, it users were able to use it for their needs, then we now probably broke their code. So, the benefit of fixing it in 0.29.x seems somewhere between zero and negative. I'll revert it and move the change to 3.0.

**Comment by navytux** [comment]
@scoder, sad to hear that.

In my view this change is absolutely a bug fix. I've tried to use `newfunc` definition, that 0.29.x privides, and hit the segfault immediately. I'm 100% sure that would be the same for anyone who tries to use `newfunc` definition from Cython/Includes.

What usually follows is that people do not bother to spend time reporting the issue to Cython upstream and just unroll their definition of newfunc in their code. Contrary to that I've tried to contribute the fix to upstream.

0.29.x is stable branch, but the change is purely a fix of bug. The patch, thus, belongs just ok to this branch. It would be a pity if 0.29.x does not have the fix but 3.0 does.

It would be also a pity if 0.29.x stops to receive fixes in the push to adopt Cython 3.0. In my view that would repeat the same sad story that Python 3 happened to go.

Kirill
